src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php

```diff
@@ -276,7 +276,7 @@ private function hasUserChanged(UserInterface $user): bool
             $userRoles[] = 'ROLE_PREVIOUS_ADMIN';
         }
 
-        if (\count($userRoles) !== \count($this->getRoleNames()) || \count($userRoles) !== \count(array_intersect($userRoles, $this->getRoleNames()))) {
+        if (\count($this->getRoleNames()) !== \count(array_intersect($this->getRoleNames(), $userRoles))) {
```

Comment:
hmm simply comparing the count of roles is not really enough, or? What if I remove one role and add another one? Then the count is still the same but the roles have changed?